src/blockchain/mod.rs

```diff
@@ -500,3 +505,197 @@ impl<K: KvStore> Blockchain for KvStoreChain<K> {
+        assert_eq!(Address::Treasury, draft.body[0].src);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_chain_should_draft_all_valid_transactions() -> Result<(), BlockchainError> {
+        let mut chain = KvStoreChain::new(db::RamKvStore::new(), Some(0x01ffff00))?;
+
+        let wallet1 = Wallet::new(Vec::from("ABC"));
+        let wallet2 = Wallet::new(Vec::from("CBA"));
+
+        let t1 = wallet1.create_transaction(wallet2.get_address(), 3, 0);
+
+        let mempool = vec![
+            t1,
+            Transaction {
+                src: Address::Treasury,
+                data: TransactionData::RegularSend { dst: wallet1.get_address(), amount: 10 },
+                nonce: 2,
+                fee: 0,
+                sig: core::Signature::Unsigned,
+            },
+        ];
+        let mut draft = chain.draft_block(1650000000, &mempool, &wallet1)?;
+
+        assert_eq!(2, draft.body.len());
+        assert_eq!(Address::Treasury, draft.body[0].src);
+
+        Ok(())
+    }
+
+
+    #[test]
+    fn test_chain_should_rollback_applied_block() -> Result<(), BlockchainError> {
+        let mut chain = KvStoreChain::new(db::RamKvStore::new(), Some(0x01ffff00))?;
+
+        let wallet1 = Wallet::new(Vec::from("ABC"));
+
+        let mut mempool = vec![
+            Transaction {
+                src: Address::Treasury,
+                data: TransactionData::RegularSend { dst: wallet1.get_address(), amount: 10 },
+                nonce: 2,
+                fee: 0,
+                sig: core::Signature::Unsigned,
+            },
+        ];
+        let mut draft = chain.draft_block(1650000000, &mempool, &wallet1)?;
+
+        draft.header.proof_of_work.nonce = 163;
+        mine_block(&chain, &mut draft)?;
+
+        chain.apply_block(&draft, false)?;
+
+        mempool.push(Transaction {
+            src: Address::Treasury,
+            data: TransactionData::RegularSend { dst: wallet1.get_address(), amount: 5 },
+            nonce: 3,
+            fee: 0,
+            sig: core::Signature::Unsigned,
+        });
+
+        let mut draft = chain.draft_block(1650000001, &mempool, &wallet1)?;
+
+        draft.header.proof_of_work.nonce = 1091;
+        mine_block(&chain, &mut draft)?;
+
+        chain.apply_block(&draft, false)?;
+
+        let height = chain.get_height()?;
+        assert_eq!(3, height);
+
+        let last_block = chain.get_block(height - 1)?;
+        assert!(matches!(&last_block.body[0].data,
+            TransactionData::RegularSend { dst: _dst_address, amount: 5 }));
+
+        chain.rollback_block()?;
+
+        let height = chain.get_height()?;
+        assert_eq!(2, height);
+
+        let account = chain.get_account(wallet1.get_address())?;
+        assert_eq!(10, account.balance);
+        // TODO: seems problematic
+        assert_eq!(0, account.nonce);
+
+        Ok(())
+    }
+
+    fn mine_block(chain: &dyn Blockchain, draft: &mut Block) -> Result<(), BlockchainError> {
```

Comment:
dont use `&dyn Blockchain`, do this instead: miner_block<B:Blockchain>(chain: &B, etc etc)